### DISCUSSION TITLE
Access Request with AuthzBackend

### DISCUSSION THREAD

**Discussion by calebbourg** [maxcountryman/axum-login#260] [Ideas]
It would be nice to have access to request data when evaluating permissions using the`AuthzBackend` interface.

An example might be something like this:

`GET /my_resources?filer_value=3`

If we wanted to check that the user requesting `my_resources` has access to the specific resources filtered by `3` we would need to have access to the query parameters of the specific request URI inside of `AuthzBackend` methods in order to evaluate.

Could achieve this by:
-  passing updating [has_perm](https://github.com/maxcountryman/axum-login/blob/9c26b37cd03be8d803ae261b7bc556229c2043da/axum-login/src/backend.rs#L183) to take an `axum::extract::Request` struct.
- updating the [macros](https://github.com/maxcountryman/axum-login/blob/9c26b37cd03be8d803ae261b7bc556229c2043da/axum-login/src/middleware.rs#L173) to provide the `Request`

This would probably be a breaking change unless there was some sane default we'd want to accept for `Request`. 
The alternative is to just write custom middleware. 

@maxcountryman if this seems like something you'd like to support, I'd be happy to open a PR.

**Comment by maxcountryman** [comment]
This system is modeled after [Django's permissions and authorization](https://docs.djangoproject.com/en/5.1/topics/auth/default/#permissions-and-authorization). Notably, Django's interface—and by extension, ours—is designed to evaluate permissions solely in relation to a user or group. The primary question we aim to answer is, "Does user A have permissions X, Y, and Z?"—independently of any additional context, such as request-specific details. This deliberate design choice ensures that permission checks are decoupled from any particular use case, fostering reusability and clarity. We want to maintain this thoughtful approach in our implementation.

Given this, it's unlikely we would modify the `AuthzBackend` interface to accept `Request` or other contextual data, as it would diverge from this fundamental principle. However, that doesn't mean request-specific permission checks are off the table.

As you pointed out, one way to handle this is by updating the macros, as they bridge the gap between general-purpose permission checks and their application in HTTP request contexts. I’m open to exploring this direction, and it would be helpful to see an example of what you have in mind for this functionality.

Alternatively, you could implement a custom macro tailored to your specific use case, which would allow you to shape things to your specific needs. That said, I do see value in making the macros more flexible relative to the request context so I think it could make sense to think more about how we might incorporate better support for that into `axum-login` directly.

**Reply by calebbourg** [reply]
Thank you for the response! 

I will put some thought into this and see if I can come up with something.

**Reply by calebbourg** [reply]
@maxcountryman here is an example of how I'm considering implementing this type of mechanism 
https://github.com/refactor-group/refactor-platform-rs/pull/78/files#diff-f1c7493b30d00626c9731e449df73adf474beaaa84fcb142cd9928c92b47b570R4-R22

which would be used in a very similar manner to the axum-login macros you've provided
Ex. https://github.com/refactor-group/refactor-platform-rs/pull/78/files#diff-58a7a3e1e2841de0eac97fdf1e09f0a1f3550532f1207443df368656fdc21c6aR186-R189


Let me know if it's something you'd like to support in axum-login and happy to open a PR.
